kv.go

```diff
@@ -10,83 +10,159 @@ import "context"
+	// Prefix iteration, only keys beginning with the designated prefix
+	// with the given prefix will be yielded.
+	//
+	// Keys exactly matching the provided `Prefix` value will not be
+	// yielded.
+	//
+	// Providing a Prefix value should cause the Start and End options
+	// to be ignored, although this is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/35
 	Prefix []byte
 
-	// Range iteration from start (inclusive) to end explusive.
-	// Start must be lexigraphically less
-	// then end, unless `nil` is used.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically greater than or
+	// equal to this value.
 	//
-	// Start defined as nil will iterate from the first
-	// key in the store.
-	// End defined as nil will iterate till the last key
-	// in the store.
+	// Providing an `End` value equal to or smaller than this value
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
 	Start []byte
-	End   []byte
 
-	// Reverse the direction of the iteration.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically smaller than this
+	// value.
+	//
+	// Providing an End value equal to or smaller than Start
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
+	End []byte
+
+	// Reverse the direction of the iteration, returning items in
+	// lexographically descending order of their keys.
 	Reverse bool
 
-	// Only iterate through keys. Calling Value() on the
-	// iterator *must* return nil (no error).
+	// Only iterate through keys. Calling Value on the
+	// iterator will return nil and no error.
+	//
+	// This option is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/34
+	//
+	// It is very likely ignored for the memory store iteration:
+	// https://github.com/sourcenetwork/corekv/issues/33
 	KeysOnly bool
 }
 
+// Reader contains read-only functions for interacting with a store.
 type Reader interface {
+	// Get returns the value at the given key.
+	//
+	// If no item with the given key is found, nil, and an [ErrNotFound]
+	// error will be returned.
 	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Has returns true if an item at the given key is found, otherwise
+	// will return false.
 	Has(ctx context.Context, key []byte) (bool, error)
+
+	// Iterator returns a read-only iterator using the given options.
 	Iterator(ctx context.Context, opts IterOptions) Iterator
 }
 
+// Writer contains functions for mutating stuff stored within a store.
 type Writer interface {
+	// Set sets the value stored against the given key.
+	//
+	// If an item already exists at the given key it will be overwritten.
 	Set(ctx context.Context, key, value []byte) error
+
+	// Delete removes the value at the given key.
+	//
+	// If no matching key is found the behaviour is undefined:
```

Comment:
suggestion: I don't think the behaviour should be undefined. A no-op is not undefined. It is a specific known behaviour.

Reply:
I don't think any of the behaviour should be undefined, that is what the linked tickets are there to resolve.

I do not know if all of the stores implementations will no-op, I think badger errors.